src/hal/src/command/mod.rs

```diff
@@ -65,100 +67,74 @@ impl Level for Primary { }
-    unsafe fn into_buffer(self) -> Cow<'a, B::CommandBuffer> { Cow::Borrowed(&self.0) }
-}
 
 /// A convenience alias for not typing out the full signature of a secondary command buffer.
-pub type SecondaryCommandBuffer<'a, B, C, S = OneShot> = CommandBuffer<'a, B, C, S, Secondary>;
+pub type SecondaryCommandBuffer<B, C, S = OneShot> = CommandBuffer<B, C, S, Secondary>;
 
 /// A strongly-typed command buffer that will only implement methods that are valid for the operations
 /// it supports.
-pub struct CommandBuffer<'a, B: Backend, C, S: Shot = OneShot, L: Level = Primary> {
-    pub(crate) raw: &'a mut B::CommandBuffer,
-    pub(crate) _marker: PhantomData<(C, S, L)>
+pub struct CommandBuffer<B: Backend, C, S = OneShot, L = Primary, R = <B as Backend>::CommandBuffer> {
+    pub(crate) raw: R,
+    pub(crate) _marker: PhantomData<(B, C, S, L)>
+}
+
+//TODO: avoid the `R` generic magic
+impl<B: Backend, C, S, L, R> Borrow<R> for CommandBuffer<B, C, S, L, R>
+where
+    B: Backend<CommandBuffer = R>,
+    R: RawCommandBuffer<B>,
+{
+    fn borrow(&self) -> &B::CommandBuffer {
+        &self.raw
+    }
 }
 
-impl<'a, B: Backend, C, S: Shot, L: Level> CommandBuffer<'a, B, C, S, L> {
+impl<B: Backend, C, S: Shot, L: Level> CommandBuffer<B, C, S, L> {
     /// Create a new typed command buffer from a raw command pool.
-    pub unsafe fn new(raw: &'a mut B::CommandBuffer) -> Self {
+    pub unsafe fn new(raw: B::CommandBuffer) -> Self {
         CommandBuffer {
-            raw: raw,
+            raw,
             _marker: PhantomData,
         }
     }
 
-    /// Get a reference to the raw command buffer
-    pub fn as_raw(&self) -> &B::CommandBuffer {
-        &*self.raw
-    }
-
-    /// Get a mutable reference to the raw command buffer
-    pub fn as_raw_mut(&mut self) -> &mut B::CommandBuffer {
-        self.raw
-    }
-
     /// Finish recording commands to the command buffers.
     ///
-    /// The command buffer will be consumed and can't be modified further.
     /// The command pool must be reset to able to re-record commands.
-    pub fn finish(self) -> Submit<B, C, S, L> {
+    pub fn finish(&mut self) {
         self.raw.finish();
-        let raw = self.raw.clone();
-
-        ::std::mem::forget(self);
+    }
 
-        Submit::new(raw)
+    /*
+    /// Get a reference to the raw command buffer
+    pub fn as_raw(&self) -> &B::CommandBuffer {
+        &self.raw
     }
 
+    /// Get a mutable reference to the raw command buffer
+    pub fn as_raw_mut(&mut self) -> &mut B::CommandBuffer {
+        &mut self.raw
+    }*/
+
     /// Downgrade a command buffer to a lesser capability type.
     ///
     /// This is safe as a downgraded version can't be `submit`'ed
     /// since `submit` requires `self` by move.
-    pub fn downgrade<D>(&mut self) -> &mut CommandBuffer<'a, B, D, S>
+    pub fn downgrade<D>(&mut self) -> &mut CommandBuffer<B, D, S>
```

Comment:
While you're at it. This method should be marked unsafe.

Reply:
Why should this method be unsafe? It explicitly says why it's safe in the doc comment.

Reply:
`std::mem::replace`